src/components/Personalization/utils/surfaceUtils.js

```diff
@@ -0,0 +1,113 @@
+import { startsWith, isNil, isNonEmptyString, includes } from "../../../utils";
+
+const SURFACE_REGEX = /^(\w+):\/\/([^/#]+)(\/[^#]*)?(#.*)?$/;
+const AUTHORITY_REGEX = /^(?:.*@)?(?:[a-z\d\u00a1-\uffff.-]+|\[[a-f\d:]+])(?::\d+)?$/;
+const PATH_REGEX = /^\/(?:[/\w\u00a1-\uffff-.~]|%[a-fA-F\d]{2})*$/;
+const FRAGMENT_REGEX = /^#(?:[/\w\u00a1-\uffff-.~]|%[a-fA-F\d]{2})+$/;
+
+const normalizePath = (path = "/") => {
+  let end = path.length;
+  while (end > 0 && "/".indexOf(path.charAt(end - 1)) !== -1) {
+    end -= 1;
+  }
+  return path.substring(0, end) || "/";
+};
+
+const parseSurface = surfaceString => {
+  const matched = surfaceString.match(SURFACE_REGEX);
+  return matched
+    ? {
+        surfaceType: isNonEmptyString(matched[1])
+          ? matched[1].toLowerCase()
+          : "",
+        authority: isNonEmptyString(matched[2]) ? matched[2].toLowerCase() : "",
+        path: isNonEmptyString(matched[3]) ? normalizePath(matched[3]) : "/",
+        fragment: matched[4]
+      }
+    : null;
+};
+
+const stringifySurface = surface =>
+  surface.surfaceType +
+  SURFACE_TYPE_DELIMITER +
+  surface.authority +
+  (surface.path || "") +
+  (surface.fragment || "");
+
+export const buildPageSurface = getPageLocation => {
+  const location = getPageLocation();
+  const host = location.host.toLowerCase();
+  const path = location.pathname;
+  return WEB + SURFACE_TYPE_DELIMITER + host + normalizePath(path);
+};
+
+const expandFragmentSurface = (surface, getPageLocation) =>
+  startsWith(surface, FRAGMENT_DELIMITER)
+    ? buildPageSurface(getPageLocation) + surface
+    : surface;
+
+const validateSurface = (surface, getPageLocation, logger) => {
+  const invalidateSurface = validationError => {
+    logger.warn(validationError);
+    return null;
+  };
+
+  if (!isNonEmptyString(surface)) {
+    return invalidateSurface(`Invalid surface: ${surface}`);
+  }
+  const expanded = expandFragmentSurface(surface, getPageLocation);
+  const parsed = parseSurface(expanded);
+  if (parsed === null) {
+    return invalidateSurface(`Invalid surface: ${surface}`);
+  }
+  if (!includes([WEB, WEBAPP], parsed.surfaceType)) {
+    return invalidateSurface(
+      `Unsupported surface type ${parsed.surfaceType} in surface: ${surface}`
+    );
+  }
+  if (!parsed.authority || !AUTHORITY_REGEX.test(parsed.authority)) {
+    return invalidateSurface(
+      `Invalid authority ${parsed.authority} in surface: ${surface}`
+    );
+  }
+  if (parsed.path && !PATH_REGEX.test(parsed.path)) {
+    return invalidateSurface(
+      `Invalid path ${parsed.path} in surface: ${surface}`
+    );
+  }
+  if (parsed.fragment && !FRAGMENT_REGEX.test(parsed.fragment)) {
+    return invalidateSurface(
+      `Invalid fragment ${parsed.fragment} in surface: ${surface}`
+    );
+  }
+  return parsed;
+};
+
+export const isPageWideSurface = scope =>
+  !!scope &&
+  scope.indexOf(WEB + SURFACE_TYPE_DELIMITER) === 0 &&
+  scope.indexOf(FRAGMENT_DELIMITER) === -1;
+
```

Comment:
I am not following, why we need this dependency injection here. Can't we send just the `surfaces` and page location and return a complex object like {surfaces, validation errors}. The caller of this function can decide to log or do something else. Otherwise `normalizeSurfaces` looks like an utility, but does a lot of stuff including logging. Testing all this via mocking etc, should be pretty painful.

Reply:
This is so we're not leaking any surface normalization-related errors to the caller. Do we really need to handle surface related valdation error logging in `createPersonalizationDetails.createQueryDetails`? At that level, we're only interested in getting valid and normalized surfaces and shouldn't be handling any validation-specific errors.
Testing is also not a problem, as we can just mock `logger.warn` as I've done in the tests